Sources/PackageLoading/ToolsVersionLoader.swift

```diff
@@ -126,59 +216,198 @@ public class ToolsVersionLoader: ToolsVersionLoaderProtocol {
+        
+        // `NSRegularExpression.firstMatch(in:options:range:)` accepts only a `String` instance as the first parameter.
+        /// The string of the Swift tools version specification.
+        let toolsVersionSpecificationString = String(toolsVersionSpecification)
+        
+        // Try to match our regex and see if the tools version specification is valid.
+        guard
+            let match = regex.firstMatch(in: toolsVersionSpecificationString, options: [], range: NSRange(toolsVersionSpecificationString.startIndex..<toolsVersionSpecificationString.endIndex, in: toolsVersionSpecificationString)),
+            // The 3 ranges are:
+            //   1. The entire matched string.
+            //   2. Capture group 1: the comment spacing.
+            //   3. Capture group 2: The version specifier.
+            // Since the version specifier is in the last range, if the number of ranges is less than 3, then no version specifier is captured by the regex.
+            // FIXME: Should this be `== 3` instead?
+            match.numberOfRanges >= 3
+        else {
+            return ManifestComponents(
+                leadingLineTerminators: leadingLineTerminators,
+                toolsVersionSpecification: toolsVersionSpecification,
+                contentsAfterToolsVersionSpecification: manifestContentsAfterToolsVersionSpecification,
+                toolsVersionSpecificationCapturedComponents: ToolsVersionSpecificationCapturedComponents(spacingAfterSlashes: nil, versionSpecifier: nil)
+            )
+        }
+        
+        // Try to match our regex and see if the tools version specification is valid.
+        guard let rangeOfSpacingAfterSlashes = Range(match.range(at: 1), in: toolsVersionSpecificationString),
+              let rangeOfVersionSpecifier = Range(match.range(at: 2), in: toolsVersionSpecificationString) else {
+            fatalError("failed to initialise instances of Range<String.Index> from valid regex match ranges")
+        }
+        
+        /// The horizontal spacing between "//" and  "swift-tools-version".
+        let spacingAfterSlashes = toolsVersionSpecificationString[rangeOfSpacingAfterSlashes]
+        /// The version number specifier.
+        let versionSpecifier = toolsVersionSpecificationString[rangeOfVersionSpecifier]
+        
+        return ManifestComponents(
+            leadingLineTerminators: leadingLineTerminators,
+            toolsVersionSpecification: toolsVersionSpecification,
+            contentsAfterToolsVersionSpecification: manifestContentsAfterToolsVersionSpecification,
+            toolsVersionSpecificationCapturedComponents: ToolsVersionSpecificationCapturedComponents(
+                spacingAfterSlashes: spacingAfterSlashes,
+                versionSpecifier: versionSpecifier
+            )
+        )
+    }
+    
+    /// The regex to match swift tools version specification.
```

Comment:
Thank you for documenting all of this!  These comments are really helpful in understanding the logic.

Reply:
I'm glad my documentation helped!